### PR TITLE
refactor: refactor environment variable handling

### PR DESCRIPTION
**Pull request by bajpai244**
Refactor environment variable parsing to a single file, and move validation to a single function.

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

As of now every time we need an environment variable, we parse it and then do a `map_err`,  in the future we might need more environment variables, and they might be used in various files { they are already being used in 2 } + some of these variables might only be needed when a special branch of the code is executed, and if not present might panic the service later.

This PR move parsing of all these variables and exposes them as lazy statics, it also provides a validation function, which can be run before the start of the service, if some critical variable is not present, it will panic the service before it even starts. Hence we can also use this function as a smoke test. i.e. making sure that we don't we do tests until this function passes.

The PR is more for our future needs, where the RPC might become more and more configurable, and for those configurations, we might need more and more environment variables.

Resolves: #199 

# What is the new behavior?

- All variables are exposed as lazy_statics 
- An environment validation function takes care that the service doesn't start before all environment variables are present
- All environment variables have been pre_fixed with `ENV` making sure that to reader it is clear that they are environment variables

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

# Other information

As of now there is only test, which is for the success case of the validation function, we can investigate in future on how to test panic cases in the same test, as panics test pass when they are run individually, but tend to fail when run with other tests.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -27,3 +27,4 @@ reth-rlp = { git = "https://github.com/paradigmxyz/reth.git", rev = "fb710e5" }
 reth-primitives = { git = "https://github.com/paradigmxyz/reth.git", rev = "fb710e5" }
 async-trait = "0.1.58"
 jsonrpsee = { version = "0.18.2", features = ["full"] }
+lazy_static = "1.4.0"
```
Comment (Eikix) [inline Cargo.toml:30]: nice refactor, good thinking

Comment (bajpai244) [reply Cargo.toml:30]: Thanks :)

File: crates/rpc/src/lib.rs [hunk 2]
```diff
@@ -22,7 +24,7 @@ pub enum RpcError {
 ///
 /// Will return `Err` if an error occurs when running the `ServerBuilder` start fails.
 pub async fn run_server(starknet_client: Box<dyn KakarotClient>) -> Result<(SocketAddr, ServerHandle), RpcError> {
-    let socket_addr = std::env::var("KAKAROT_HTTP_RPC_ADDRESS").unwrap_or_else(|_| "0.0.0.0:3030".to_owned());
+    let socket_addr = ENV_KAKAROT_HTTP_RPC_ADDRESS.clone();
```
Comment (Eikix) [inline crates/rpc/src/lib.rs:27]: since this is used in run_server and not main, I think it should be separate

**Review by Eikix — CHANGES_REQUESTED** [review]
As discussed in Daily, let's try to change the approach!:)

File: crates/core/src/client/config.rs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    pub kakarot_address: FieldElement,
+    pub proxy_account_class_hash: FieldElement,
+}
+
+impl StarknetConfig {
+    pub fn new(
+        starknet_rpc: &str,
+        kakarot_address: FieldElement,
+        proxy_account_class_hash: FieldElement,
+    ) -> StarknetConfig {
```
Comment (Eikix) [inline crates/core/src/client/config.rs:19]: this can be `self`
funny that clippy didnt catch it

Comment (bajpai244) [reply crates/core/src/client/config.rs:19]: Oh! Makes sense!

Yeah, something that should have been caught by Clippy!

File: crates/core/src/models/cfg.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
+use std::panic;
```
Comment (Eikix) [inline crates/core/src/models/cfg.rs:1]: either this or `config.rs` are duplicated, you should probably choose one
additionally, we should try not to use shortened names cfg -> config and StarknetCfg -> StarknetConfig

File: crates/core/src/models/cfg.rs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    Ok(StarknetCfg {
+            starknet_rpc: rpc_url,
+            kakarot_address,
+            proxy_account_class_hash
+        })
+    }
+}
+
+
+// validate_env makes sure that are all required environment variables are set and have valid values.
```
Comment (Eikix) [inline crates/core/src/models/cfg.rs:40]: prefer rustdoc style comments

File: crates/core/src/client/config.rs [hunk 6]
```diff
@@ -0,0 +1,37 @@
+use eyre::{eyre, Result};
+use starknet::core::types::FieldElement;
+
+fn get_env_var(name: &str) -> Result<String> {
```
Comment (Eikix) [inline crates/core/src/client/config.rs:4]: Based on @greged93 's former PRs we could use `thiserror` here to define a configerror, and then add it in the more general error enum as a transparent error

**Review by Eikix — CHANGES_REQUESTED** [review]
this is going in the right direction! ggs

File: crates/core/src/models/cfg.rs (hunk 4 above)
Comment (bajpai244) [reply crates/core/src/models/cfg.rs]: removed it.

File: crates/rpc/Cargo.toml [hunk 7]
```diff
@@ -36,6 +36,7 @@ starknet = { workspace = true }
 thiserror = "1.0.38"
 tracing = "0.1.34"
 tracing-subscriber = { version = "0.3.3", features = ["env-filter"] }
+lazy_static = { workspace = true }
```
Comment (Eikix) [inline crates/rpc/Cargo.toml:39]: do we still need that?

**Comment by Eikix** [comment]
Careful, big rebase;p

**Comment by bajpai244** [comment]
> Careful, big rebase;p

XD

File: crates/rpc/Cargo.toml (hunk 7 above)
Comment (bajpai244) [reply crates/rpc/Cargo.toml:39]: We can move it back from a workspace dependency to a crate specific dependency.

**Review by Eikix — DISMISSED** [review]
this is great! well done! clean and looking great! 


side quest: optionally, before merge, we could rename all `*_cfg` to `*_config` as I usually prefer non-abbreviated names for variables (e.g. `starknet_config`, `rpc_config`)
to be fair, this is not very impactful, i'm aware that rust uses cfg(test) and many other abbreviations so it's fine,

it's simply that in the kakarot-rpc codebase, i notice we either use full names or initials (e.g. rpc) but not abbreviations

**Comment by bajpai244** [comment]
> this is great! well done! clean and looking great!
> 
> side quest: optionally, before merge, we could rename all `*_cfg` to `*_config` as I usually prefer non-abbreviated names for variables (e.g. `starknet_config`, `rpc_config`) to be fair, this is not very impactful, i'm aware that rust uses cfg(test) and many other abbreviations so it's fine,
> 
> it's simply that in the kakarot-rpc codebase, i notice we either use full names or initials (e.g. rpc) but not abbreviations

It makes sense and we should be consistent :)

Changed and pushed!

**Review by Eikix — APPROVED** [review]
great job

**Pull request merged**